contributors/design-proposals/defer-containers.md

```diff
@@ -0,0 +1,248 @@
+# Defer Containers
+
+dhilip.kumar.s@huawei.com
+
+March 2017
+
+## Prerequisite
+This is a continuation of [initContainers](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/container-init.md)
+
+## Motivation
+Since the introduction to Statefulset and Daemonset PODs that consist of stateful containers are now fully supported by
+Kubernetes.  Such stateful pods benefit from constructor and destructor semantics. InitContainers provide constructor
+semantics. The proposed DeferContainers will provide destructor semantics for such stateful PODs
+
+## Abstract
+This introduces the concept of ‘deferContainers’, inspired by golang’s keyword ‘defer’. This will allow one to define a
+set of Containers to be executed at the time of POD termination. The defined containers will be executed sequentially and
+in the specified order. It will bring destructor() capability to Pods.
+
+## Use Cases
+
+### Cleanup:
```

Comment:
We'd have to come up with some other signal.  --grace-period=1 means "delete fast" and --grace-period=0 means "give up strong consistency and risk split brain".  You don't want to do the latter.

I'm not sure there is a backward compatible way to allow defer containers to run without being bounded in time by the grace-period.  I'm also not sure we want defer containers to run unbounded, since that was a key part of the graceful deletion design.